src/project/select-modules-in-sources.ts

```diff
@@ -0,0 +1,342 @@
+        },
+    });
+}
+
+interface SelectModulesInProjectSourceParams {
+    readonly source: ProjectSource;
+    readonly options: ProjectOptions;
+    readonly modelWithModules: Model;
+    readonly selectedModules: Set<string>;
+    readonly validationContext: ValidationContext;
+    readonly removeModuleDeclarations: boolean;
+}
+
+function selectModulesInProjectSource(
+    params: SelectModulesInProjectSourceParams,
+): ProjectSource | undefined {
+    const parsedSource = parseProjectSource(
+        params.source,
+        params.options,
+        params.validationContext,
+    );
+    if (!parsedSource) {
+        return undefined;
+    }
+    let newBody;
+    switch (parsedSource.kind) {
+        case ParsedProjectSourceBaseKind.OBJECT:
+            newBody = selectModulesInObjectSource({ parsedSource, ...params });
+            break;
+        case ParsedProjectSourceBaseKind.GRAPHQL:
+            newBody = selectModulesInGraphQLSource({ parsedSource, ...params });
+            break;
+    }
+    if (newBody === undefined) {
+        return undefined;
+    } else if (newBody === params.source.body) {
+        return params.source;
+    } else {
+        return new ProjectSource(params.source.name, newBody, params.source.filePath);
+    }
+}
+
+interface SelectModulesInObjectSourceParams extends SelectModulesInProjectSourceParams {
+    readonly parsedSource: ParsedObjectProjectSource;
+}
+
+function selectModulesInObjectSource({
+    source,
+    parsedSource,
+    removeModuleDeclarations,
+    selectedModules,
+}: SelectModulesInObjectSourceParams): string | undefined {
+    if ('modules' in parsedSource.object) {
+        // if the file *only* includes module declarations, and we should remove them, we can just remove the file
+        if (removeModuleDeclarations && Object.keys(parsedSource.object).length === 1) {
+            return undefined;
+        }
+
+        // TODO improve editing object sources
+        // we need to remove the modules but keep the rest
+        // canonot just remove the spans because yaml is whitespace-sensitive
+        // should maybe use a source-preserving yaml editor (as yaml is a json superset)
+        // for now, just remove the formatting
+        const newObject = { ...parsedSource.object };
+        if (removeModuleDeclarations) {
+            delete newObject.modules;
+        } else if (Array.isArray(parsedSource.object.modules)) {
+            // if we shouldn't remove the module declarations completely, filter it down
+            newObject.modules = parsedSource.object.modules.filter((m) => selectedModules.has(m));
+        }
+        return JSON.stringify(newObject, undefined, '  ');
+    }
+
+    return source.body;
+}
+
+interface SelectModulesInGraphQLSourceParams extends SelectModulesInProjectSourceParams {
+    readonly parsedSource: ParsedGraphQLProjectSource;
+}
+
+interface Change {
+    readonly oldLocation: Location;
```

Comment:
Why is it called `oldLocation`? I think just calling it `location` would be less confusing

Reply:
hm true, I don't remember choosing that name :D